modules/sitemaps/sitemaps.php

```diff
@@ -1,792 +1,1865 @@
+
+		$log->report( 'Page Sitemaps:' );
+
+		// Generate sitemaps until no posts remain.
+		while ( true === $any_posts_left ) {
+			$result = $this->build_one_page_sitemap(
+				$sitemap_number,
+				$post_id,
+				$log
+			);
 
-			if ( ! empty( $images ) ) {
-				$url['image:image'] = $images;
+			if ( true === $result['any_posts_left'] ) {
+				$post_id = $result['last_post_id'];
+				$sitemap_number += 1;
+			} else {
+				$any_posts_left = false;
 			}
 		}
 
-		if ( $post->post_modified_gmt && $post->post_modified_gmt != '0000-00-00 00:00:00' ) {
-			$post_latest_mod = $post->post_modified_gmt;
+		// Clean up old page sitemaps.
+		$log->report( '-- Cleaning up old page sitemaps.' );
+		$this->delete_numbered_posts_after(
+			'sitemap-',
+			$sitemap_number,
+			'jp_sitemap'
+		);
+
+		// Return the number of the last sitemap to be stored.
+		return $sitemap_number;
+	}
+
+	/**
+	 * Build and store a single page sitemap.
+	 *
+	 * Side effect: Create/update a jp_sitemap post.
+	 *
+	 * @access private
+	 * @since 4.5.0
+	 *
+	 * @param int                    $number The number of the current sitemap.
+	 * @param int                    $from_id The greatest lower bound of the IDs of the posts to be included.
+	 * @param Jetpack_Sitemap_Logger $log Logger object for reporting debug messages.
+	 *
+	 * @return array @args {
+	 *   @type int $last_post_id The ID of the last post to be successfully added to the buffer.
+	 *   @type bool $any_posts_left 'true' if there are posts which haven't been saved to a sitemap, 'false' otherwise.
+	 * }
+	 */
+	private function build_one_page_sitemap( $number, $from_id, $log ) {
+		$last_post_id = $from_id;
+		$any_posts_left = true;
+
+		$log->report( '-- Building Page Sitemap ' . $number . '.' );
+
+		$path_url = $this->the_jetpack_sitemap_uri();
+
+		$jetpack_version = JETPACK__VERSION;
+
+		$buffer = new Jetpack_Sitemap_Buffer(
+			self::SITEMAP_MAX_ITEMS,
+			self::SITEMAP_MAX_BYTES,
+			<<<HEADER
+<?xml version='1.0' encoding='UTF-8'?>
+<!-- generator='jetpack-{$jetpack_version}' -->
+<?xml-stylesheet type='text/xsl' href='{$path_url}/sitemap.xsl'?>
+<urlset xmlns='http://www.sitemaps.org/schemas/sitemap/0.9'>\n
+HEADER
+			,
+			<<<FOOTER
+</urlset>\n
+FOOTER
+			,
+			/* epoch */
+			'1970-01-01 00:00:00'
+		);
+
+		// Add entry for the main page (only if we're at the first one).
+		if ( 1 === $number ) {
+			$home_url = home_url();
+			$buffer->try_to_add_item(<<<XML
+<url>
+ <loc>{$home_url}</loc>
+</url>\n
+XML
+			);
 		}
-		if ( $post->comment_count > 0 ) {
-			// last modified based on last comment
-			$latest_comment_datetime = jetpack_get_approved_comments_max_datetime( $post->ID );
-			if ( ! empty( $latest_comment_datetime ) ) {
-				if ( is_null( $post_latest_mod ) || $latest_comment_datetime > $post_latest_mod ) {
-					$post_latest_mod = $latest_comment_datetime;
+
+		while ( false === $buffer->is_full() ) {
+			$posts = $this->query_posts_after_id( $last_post_id, 1000, $log );
```

Comment:
I fixed this by making the logger object a class member, rather than passing one around.